// web3-plugin-zksync/lib/web3.js/packages/web3-types/src/web3_base_provider.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Web3BaseProvider = void 0;
const symbol = Symbol.for('web3/base-provider');
class Web3BaseProvider {
    static isWeb3Provider(provider) {
        return (provider instanceof Web3BaseProvider ||
            Boolean(provider && provider[symbol]));
    }
    get [symbol]() {
        return true;
    }
    send(payload, callback) {
        this.request(payload)
            .then(response => {
            callback(null, response);
        })
            .catch((err) => {
            callback(err);
        });
    }
    async sendAsync(payload) {
        return this.request(payload);
    }
    asEIP1193Provider() {
        const newObj = Object.create(this);
        const originalRequest = newObj.request;
        newObj.request = async function request(args) {
            const response = (await originalRequest(args));
            return response.result;
        };
        newObj.asEIP1193Provider = undefined;
        return newObj;
    }
}
exports.Web3BaseProvider = Web3BaseProvider;
