// web3-plugin-zksync/lib/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MAX_INTEGER = exports.EIP712_TYPES = exports.REQUIRED_L1_TO_L2_GAS_PER_PUBDATA_LIMIT = exports.DEFAULT_GAS_PER_PUBDATA_LIMIT = exports.L1_RECOMMENDED_MIN_ETH_DEPOSIT_GAS_LIMIT = exports.L1_RECOMMENDED_MIN_ERC20_DEPOSIT_GAS_LIMIT = exports.L1_FEE_ESTIMATION_COEF_DENOMINATOR = exports.L1_FEE_ESTIMATION_COEF_NUMERATOR = exports.MAX_BYTECODE_LEN_BYTES = exports.PRIORITY_OPERATION_L2_TX_TYPE = exports.EIP712_TX_TYPE = exports.EIP1271_MAGIC_VALUE = exports.L1_TO_L2_ALIAS_OFFSET = exports.NONCE_HOLDER_ADDRESS = exports.L2_BASE_TOKEN_ADDRESS = exports.L2_ETH_TOKEN_ADDRESS = exports.L1_MESSENGER_ADDRESS = exports.CONTRACT_DEPLOYER_ADDRESS = exports.BOOTLOADER_FORMAL_ADDRESS = exports.ETH_ADDRESS_IN_CONTRACTS = exports.LEGACY_ETH_ADDRESS = exports.ETH_ADDRESS = exports.ZERO_HASH = exports.ZERO_ADDRESS = void 0;
exports.ZERO_ADDRESS = '0x0000000000000000000000000000000000000000';
exports.ZERO_HASH = '0x0000000000000000000000000000000000000000000000000000000000000000';
exports.ETH_ADDRESS = '0x0000000000000000000000000000000000000000';
exports.LEGACY_ETH_ADDRESS = '0x0000000000000000000000000000000000000000';
exports.ETH_ADDRESS_IN_CONTRACTS = '0x0000000000000000000000000000000000000001';
exports.BOOTLOADER_FORMAL_ADDRESS = '0x0000000000000000000000000000000000008001';
exports.CONTRACT_DEPLOYER_ADDRESS = '0x0000000000000000000000000000000000008006';
exports.L1_MESSENGER_ADDRESS = '0x0000000000000000000000000000000000008008';
exports.L2_ETH_TOKEN_ADDRESS = '0x000000000000000000000000000000000000800a';
exports.L2_BASE_TOKEN_ADDRESS = '0x000000000000000000000000000000000000800a';
exports.NONCE_HOLDER_ADDRESS = '0x0000000000000000000000000000000000008003';
exports.L1_TO_L2_ALIAS_OFFSET = '0x1111000000000000000000000000000000001111';
exports.EIP1271_MAGIC_VALUE = '0x1626ba7e';
exports.EIP712_TX_TYPE = 0x71;
exports.PRIORITY_OPERATION_L2_TX_TYPE = 0xff;
exports.MAX_BYTECODE_LEN_BYTES = ((1 << 16) - 1) * 32;
exports.L1_FEE_ESTIMATION_COEF_NUMERATOR = 12;
exports.L1_FEE_ESTIMATION_COEF_DENOMINATOR = 10;
exports.L1_RECOMMENDED_MIN_ERC20_DEPOSIT_GAS_LIMIT = 400000;
exports.L1_RECOMMENDED_MIN_ETH_DEPOSIT_GAS_LIMIT = 200000;
exports.DEFAULT_GAS_PER_PUBDATA_LIMIT = 50000;
exports.REQUIRED_L1_TO_L2_GAS_PER_PUBDATA_LIMIT = 800;
exports.EIP712_TYPES = {
    EIP712Domain: [
        { name: 'name', type: 'string' },
        { name: 'version', type: 'string' },
        { name: 'chainId', type: 'uint256' },
    ],
    Transaction: [
        { name: 'txType', type: 'uint256' },
        { name: 'from', type: 'uint256' },
        { name: 'to', type: 'uint256' },
        { name: 'gasLimit', type: 'uint256' },
        { name: 'gasPerPubdataByteLimit', type: 'uint256' },
        { name: 'maxFeePerGas', type: 'uint256' },
        { name: 'maxPriorityFeePerGas', type: 'uint256' },
        { name: 'paymaster', type: 'uint256' },
        { name: 'nonce', type: 'uint256' },
        { name: 'value', type: 'uint256' },
        { name: 'data', type: 'bytes' },
        { name: 'factoryDeps', type: 'bytes32[]' },
        { name: 'paymasterInput', type: 'bytes' },
    ],
};
exports.MAX_INTEGER = BigInt('0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff');
