// web3-plugin-zksync/lib/paymaster-utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PAYMASTER_FLOW_ABI = void 0;
exports.getApprovalBasedPaymasterInput = getApprovalBasedPaymasterInput;
exports.getGeneralPaymasterInput = getGeneralPaymasterInput;
exports.getPaymasterParams = getPaymasterParams;
const web3_eth_contract_1 = require("web3-eth-contract");
const IPaymasterFlow_1 = require("./contracts/IPaymasterFlow");
exports.PAYMASTER_FLOW_ABI = new web3_eth_contract_1.Contract(IPaymasterFlow_1.Abi);
function getApprovalBasedPaymasterInput(paymasterInput) {
    return exports.PAYMASTER_FLOW_ABI.methods
        .approvalBased(paymasterInput.token, paymasterInput.minimalAllowance, paymasterInput.innerInput)
        .encodeABI();
}
function getGeneralPaymasterInput(paymasterInput) {
    return exports.PAYMASTER_FLOW_ABI.methods.general(paymasterInput.innerInput).encodeABI();
}
function getPaymasterParams(paymasterAddress, paymasterInput) {
    if (paymasterInput.type === 'General') {
        return {
            paymaster: paymasterAddress,
            paymasterInput: getGeneralPaymasterInput(paymasterInput),
        };
    }
    else {
        return {
            paymaster: paymasterAddress,
            paymasterInput: getApprovalBasedPaymasterInput(paymasterInput),
        };
    }
}
