// web3-plugin-zksync/lib/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AccountNonceOrdering = exports.AccountAbstractionVersion = exports.TransactionStatus = exports.PriorityOpTree = exports.PriorityQueueType = exports.Network = exports.ZeroHash = exports.ZeroAddress = void 0;
exports.ZeroAddress = '0x0000000000000000000000000000000000000000';
exports.ZeroHash = '0x0000000000000000000000000000000000000000000000000000000000000000';
var Network;
(function (Network) {
    Network[Network["Mainnet"] = 1] = "Mainnet";
    Network[Network["Ropsten"] = 3] = "Ropsten";
    Network[Network["Rinkeby"] = 4] = "Rinkeby";
    Network[Network["Goerli"] = 5] = "Goerli";
    Network[Network["Sepolia"] = 6] = "Sepolia";
    Network[Network["Localhost"] = 9] = "Localhost";
    Network[Network["EraTestNode"] = 10] = "EraTestNode";
})(Network || (exports.Network = Network = {}));
var PriorityQueueType;
(function (PriorityQueueType) {
    PriorityQueueType[PriorityQueueType["Deque"] = 0] = "Deque";
    PriorityQueueType[PriorityQueueType["HeapBuffer"] = 1] = "HeapBuffer";
    PriorityQueueType[PriorityQueueType["Heap"] = 2] = "Heap";
})(PriorityQueueType || (exports.PriorityQueueType = PriorityQueueType = {}));
var PriorityOpTree;
(function (PriorityOpTree) {
    PriorityOpTree[PriorityOpTree["Full"] = 0] = "Full";
    PriorityOpTree[PriorityOpTree["Rollup"] = 1] = "Rollup";
})(PriorityOpTree || (exports.PriorityOpTree = PriorityOpTree = {}));
var TransactionStatus;
(function (TransactionStatus) {
    TransactionStatus["NotFound"] = "not-found";
    TransactionStatus["Processing"] = "processing";
    TransactionStatus["Committed"] = "committed";
    TransactionStatus["Finalized"] = "finalized";
})(TransactionStatus || (exports.TransactionStatus = TransactionStatus = {}));
var AccountAbstractionVersion;
(function (AccountAbstractionVersion) {
    AccountAbstractionVersion[AccountAbstractionVersion["None"] = 0] = "None";
    AccountAbstractionVersion[AccountAbstractionVersion["Version1"] = 1] = "Version1";
})(AccountAbstractionVersion || (exports.AccountAbstractionVersion = AccountAbstractionVersion = {}));
var AccountNonceOrdering;
(function (AccountNonceOrdering) {
    AccountNonceOrdering[AccountNonceOrdering["Sequential"] = 0] = "Sequential";
    AccountNonceOrdering[AccountNonceOrdering["Arbitrary"] = 1] = "Arbitrary";
})(AccountNonceOrdering || (exports.AccountNonceOrdering = AccountNonceOrdering = {}));
