// web3-plugin-zksync/lib/web3-plugin-zksync/src/paymaster-utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getPaymasterParams = exports.getGeneralPaymasterInput = exports.getApprovalBasedPaymasterInput = exports.PAYMASTER_FLOW_ABI = void 0;
const web3_eth_contract_1 = require("web3-eth-contract");
const IPaymasterFlow_1 = require("./contracts/IPaymasterFlow");
exports.PAYMASTER_FLOW_ABI = new web3_eth_contract_1.Contract(IPaymasterFlow_1.Abi);
function getApprovalBasedPaymasterInput(paymasterInput) {
    return exports.PAYMASTER_FLOW_ABI.methods
        .approvalBased(paymasterInput.token, paymasterInput.minimalAllowance, paymasterInput.innerInput)
        .encodeABI();
}
exports.getApprovalBasedPaymasterInput = getApprovalBasedPaymasterInput;
function getGeneralPaymasterInput(paymasterInput) {
    return exports.PAYMASTER_FLOW_ABI.methods.general(paymasterInput.innerInput).encodeABI();
}
exports.getGeneralPaymasterInput = getGeneralPaymasterInput;
function getPaymasterParams(paymasterAddress, paymasterInput) {
    if (paymasterInput.type === 'General') {
        return {
            paymaster: paymasterAddress,
            paymasterInput: getGeneralPaymasterInput(paymasterInput),
        };
    }
    else {
        return {
            paymaster: paymasterAddress,
            paymasterInput: getApprovalBasedPaymasterInput(paymasterInput),
        };
    }
}
exports.getPaymasterParams = getPaymasterParams;
